### DISCUSSION TITLE
Validations on components

### DISCUSSION THREAD

**Discussion by RezaKargar** [bitfoundation/bitplatform#1666] [Q&A] [answered]
Hi to all 👋

I just opened this discussion to have some sort of Q&A discussion 😁🤷‍♂️

<br/>

I saw some issues about implementing the validations for some components,
I read the similar solved issues code changes and got some kind of understanding the way how to implement the validation for components
I just mention the steps that I think should be followed to solve those kind of issues, please check them and correct me it they're incorrect:

1. Change the component to inherit `BitComponentBase<T>` instead of `BitComponentBase`
2. Take care of conflicts for `Value` property of **BitComponentBase** and rename the actual **value** property of component itself if needed
3. Override the `TryParseValueFromString` method of **BitComponentBase**
    * I think this is for trying to pars the value of component from a string and if the parsing fails, give corresponding validation message otherwise result will be the parsed value 
4. Modify `RegisterComponentClasses` method for providing the **-invalid-** style class when value is invalid
5. Add corresponding **-invalid-** css style for the component

<br/><br/>

I know this may be long but there are some more unknowns for me:
- How can I make the component to be invalid for test purposes?
- When I should implement and override `Dispose` method and when not?

<br/>

I'll be so grateful if you help me on this,
I try to learn from your repo and also try to help ✌️😉

**Comment by msynk** [comment, marked as answer]
Hello Reza,

thanks for submitting this discussion and introducing this important topic.

first of all only the input components can have validation infrastructure in them.
input components are all components that in a way are receiving a specific value from the user interaction.

and secondly, the base class should be `BitInputBase<T>` type.

other parts of your demonstration are correct.

in order to test the validation, you need a separate test `.razor` file with a specific `EditForm` component in it and a model with .net validation enabled (through DataAnnotation attributes or any other ways).
(we have a sample coming up in a PR that is under review atm)

The dispose method should be used like any other classes in .net, whenever you need to dispose of something 😉
